### DISCUSSION TITLE
Will there be LiveView / WebSocket support in the future?

### DISCUSSION THREAD

**Discussion by lushmana** [leptos-rs/leptos#170] [Ideas]
Leptos, looks like the next generation of Elm based web development using Rust, having spent the last couple of years doing Elixir Phoenix LiveView, is there anything on the Roadmap that looks like LiveView, which by the way uses an integrated WebSocket, around an Actor model (GenServer), with a native templating engine all authored in Elixr. 
As there are many Web development tools for Rust, I see one advantage of using Rust is to produce the fastest compiled output, but productivity through a clean coding experience is critical for the uptake.

**Comment by gbj** [comment]
Not really. This rendering model is quite different from Leptos's, which is built on fine-grained reactivity. I know Dioxus has some LiveView features either already or upcoming; you may want to check them out!

Leptos is really *very* far from Elm in its approach. There are plenty of Elm-ish Rust frameworks like Yew, Sauron, or Seed, which may be interesting to you too.

**Reply by lushmana** [reply]
What makes Leptos interesting is that the client code is still rust (WebAssembly) NOT js and Server Code of Rust as well, that makes it similar to Elixir/Phoenix, if the fine-grained reactivity achieves similar performance to LiveView, then developers won't really care. To me it's developing in the one language, for both the Server and Client that is appealing.

**Reply by gbj** [reply]
Oh, well in that case — yes, we have that! If you want a good example of how to write very tightly coupled client and server code check out the [todo-app-sqlite](https://github.com/gbj/leptos/tree/main/examples/todo-app-sqlite) example. You really can basically write the application as though the client/server distinction wasn't that big of a deal.

**Comment by d4h0** [comment]
> Not really. This rendering model is quite different from Leptos's, which is built on fine-grained reactivity.

@gbj: Wouldn't fine-grained reactivity be perfectly suited for an implementation of LiveView?

Instead of applying the changes directly to the DOM, you'd just sent them from the server (which manages the state of the app and handles user events) to the browser (which sends events to the server).

LiveView, basically, is just a layer between the DOM and a VDOM or the fine-grained reactivity of Leptos.

I might miss something, but fine-grained reactivity seems perfect for this.

**Reply by gbj** [reply]
Sure, you could do that, the question is just: Why?

By which I mean: it's perfectly possible to implement this in userland for any given application or as a library. For particular kinds of applications like games, chat rooms, or collaborative editing it might be useful. As a general purpose rendering technique it's not the most straightforward or performant option.

**Comment by tv42** [comment]
What liveview style can do, that gets really messy with "server functions", is incremental, automatic changes as server-side data changes. For example, the todo example has to manually list all the operations that are known to invalidate the previous RPC, and even then does not see changes made by other clients.

You can build something like this on top of SSE or WebSockets yourself, essentially subscribing the client to server-side pubsub events, but now you're building layers upon layers of communication mechanisms -- with liveview, the liveview itself is the pubsub mechanism, and all the reactive updates are handled the same way, requiring less application programmer effort.

https://github.
com/leptos-rs/leptos/blob/274fe07daeea60077aabbf76fd7eb44ed725f328/examples/todo_app_sqlite_axum/src/todo.rs#L123-L133

I'm currently experimenting with dioxus-liveview and they are talking of abandoning it, and it does not handle disconnects right. I wish there was a standard "shove (V)DOM mutations over the wire" convention I could plug any renderer into..

**Reply by gbj** [reply]
> You can build something like this on top of SSE or WebSockets yourself, essentially subscribing the client to server-side pubsub events

Correct. From the very earliest days of the framework, we've had an example of this in `examples/counter_isomorphic` using SSE. More recently, there's added support for websockets created like server functions, which you can see in the `examples/websockets` example.

> but now you're building layers upon layers of communication mechanisms

Well, a layer upon another layer, in any case. 

It's pretty straightforward to build things like collaborative editing over websockets -- either by doing it yourself or by using one of the various databases that supports it. The hard parts are the parts that have the least to do with the UI (how you manage conflicting changes, etc.)

**Comment by kinthaiofficial** [comment]
LiveView/WebSocket support in Leptos would make it a much stronger choice for real-time UIs where you want server-driven updates without client-side state management complexity.

**The Phoenix LiveView model (which Leptos is likely drawing inspiration from):**
- UI state lives on the server as a process (or struct in Rust's case)
- Client sends user events (clicks, input changes) → server updates state → server pushes minimal diffs to client
- No client-side JavaScript state to sync — the server is the source of truth

**What this requires technically from Leptos:**
1. **Persistent WebSocket connection per component tree** — the SSR'd component must maintain a connection to its server-side counterpart
2. **Efficient diffing** — when server state changes, compute the minimal DOM diff and send only that. Sending the full HTML on each update defeats the purpose.
3. **Server-side component lifecycle** — components need a clear lifecycle tied to the WebSocket connection lifespan, not the HTTP request
4. **Backpressure** — if the client can't keep up with server updates (slow connection, expensive renders), the server needs to be able to skip or batch intermediate states

**Why this is hard for Rust:**
Phoenix LiveView works because Elixir's actor model makes "one process per session" cheap. In Rust, you need async tasks serving the same role, which works but requires careful resource management — especially at scale where thousands of concurrent LiveView connections are running.

The Axum integration + tokio actor model could make this tractable. Is there an RFC or issue tracking this feature's design?

**Reply by gbj** [reply]
This has been discussed adequately in the discussion above. Please read my replies there before posting an Claude-generated response. (I'm just assuming it's Claude based on its idioms, but whatever AI tool you're using does not understand what Leptos is *at all*.)
